### DISCUSSION TITLE
How to disable IP blocking for many requests

### DISCUSSION THREAD

**Discussion by NeedHelp4Indy** [IndySockets/Indy#585] [Q&A]
Hello all,

I have the following problem: My delphi application blocks an entire IP address range for a certain number of requests. After a manual restart of the process everything works as expected. Is there a way to stop the blocking of IP ranges? Or alternatively to increase the limit?
If yes, would be great if someone can share his knowledge.

Kind regards and best thanks!

**Comment by rlebeau** [comment]
I don't understand what you are describing.  Which Indy component(s) are you using?  What exactly is being blocked?  Is the blockage in your app, or in other apps?  Can you please clarify the issue with more detail?  Thanks.

**Comment by NeedHelp4Indy** [comment]
The blockage is in our app. If there are too many HTTP requests in a short time window (from another server), it seems that our app activate a security mechanism. There must be some kind of blacklist in which requests from the subnets are blocked directly (from which the high load originated).
If you try to reach the app from another network, everything works as it should.
The blacklist can currently be refreshed by restarting the application. However, this is only a workaround in my eyes and I am looking to see if there are potentially other solutions.

Which Indy components are used exactly, I don't know myself at the moment. If absolutely necessary, I would try to list them.

Thanks in advance & Best regards!

**Comment by rlebeau** [comment]
> The blockage is in our app. If there are too many HTTP requests in a short time window (from another server), it seems that our app activate a security mechanism.

What EXACTLY are you seeing?  Can you provide a log or screenshot?  How short of a window, and how many requests in that window?

> There must be some kind of blacklist in which requests from the subnets are blocked directly (from which the high load originated).

If there is, that is not in Indy itself, that would have to be in either an OS firewall or on the network itself.

> The blacklist can currently be refreshed by restarting the application. However, this is only a workaround in my eyes and I am looking to see if there are potentially other solutions.

Not without first understanding what is actually happening to begin with.

**Comment by NeedHelp4Indy** [comment]
At a certain point, HTTP requests are no longer accepted. From the client's point of view, it seems that the server on which the application is running is unreachable.

However, requests from any other client are processed correctly.

I can possibly provide the following information about the components:
A CustomHTTPServer is used. I suspect that the error is related to the "SessionList" in some way.
![image](https://user-images.githubusercontent.com/131161379/237029452-2a7b3d4b-b5b7-4101-9e2c-29586e9b7a4b.png)



I have just tried different things, so far without success.
Do you have any further experience here, is it because of the SessionList or are there other error sources?

**Comment by NeedHelp4Indy** [comment]
Additional Information: it seems that the problem is only with HTTPS communication, not with HTTP.

**Comment by rlebeau** [comment]
> At a certain point, HTTP requests are no longer accepted. From the client's point of view, it seems that the server on which the application is running is unreachable.

That is likely, if the network (not Indy) is blocking the connections. But again, I would need to see EXACTLY what you are seeing, I can't diagnose your issue with the information you have provided so far.  What is the EXACT error you are seeing?

> A CustomHTTPServer is used. I suspect that the error is related to the "SessionList" in some way.

I don't see how it could be.  The `SessionList` is merely a cache of `TIdHTTPSession` objects in memory so they can be shared from one HTTP request to another.  It has nothing to do with the server's ability to accept client connections or HTTP requests.  When a new session is created, a cookie is sent back to the client, which can then be sent back to the server on subsequent requests.  If a request is received without the cookie, or if the specified session has expired, then a new session is created.

> Do you have any further experience here, is it because of the SessionList or are there other error sources?

It would have to be caused by an external source, it is not Indy itself blocking the connections.

> Additional Information: it seems that the problem is only with HTTPS communication, not with HTTP.

Have you tried using a packet sniffer like Wireshark to look at the TLS handshakes on each connection?  Are they being responded to?  Are they being rejected?

**Comment by NeedHelp4Indy** [comment]
If requests are sent from the client, the behavior looks like this:

![image](https://github.com/IndySockets/Indy/assets/131161379/46e73ec1-377c-45cb-84bc-36130b449d1c)

The behavior is due to the TLS handshake responses being reset immediately.
The handshakes are answered, but reset instantly. I was able to detect this with Wireshark.#

![MicrosoftTeams-image](https://github.com/IndySockets/Indy/assets/131161379/7ce5aaa6-cf0d-417f-931f-152cce632087)


There are no error messages on the server and in our applications.

Now if requests are sent from another client, the application responds. From the client, which is misinterpreted in some way as a "DOS attack", all requests are blocked until OUR application (not server!) is restarted. This is the reason I am having such a hard time describing the behavior I am seeing.  

Maybe it is also blocked on OS level. Unfortunately, I still could not fix the error. Alternatively, we can try to find a joint appointment, then I could show them the problem live.

**Comment by rlebeau** [comment]
> The behavior is due to the TLS handshake responses being reset immediately. The handshakes are answered, but reset instantly. I was able to detect this with Wireshark.#

I see nothing in that screenshot that implies TLS is involved.  The TCP handshake is being rejected immediately, before the TLS handshake has a chance to begin.  Most likely, a firewall/router sitting between your client and the server has blacklisted your client IP from making new connections to the server.  Again, nothing related to Indy itself.

> There are no error messages on the server

Right, because the server is not even seeing the TCP connection to begin with.  The network/OS is rejecting the connection, so the server's listening socket doesn't see it and can't accept it.

> Now if requests are sent from another client, the application responds. From the client, which is misinterpreted in some way as a "DOS attack", all requests are blocked until OUR application (not server!) is restarted.

Is your client app binding to the same local IP/Port for each new outgoing connection it makes?  Or, is it using a random Port each time (as it should be)?  In the latter case, since the blocked connections are coming from the same process, that would imply that the blocker is the OS/firewall on the same local machine that the client app is running on, since the PID changes if the app is closed and restarted.  A remote blocker would not be able to identify new processes.

> Maybe it is also blocked on OS level. Unfortunately, I still could not fix the error. Alternatively, we can try to find a joint appointment, then I could show them the problem live.

You will likely have to either 1) make your client app make fewer requests in a short time, or 2) whitelist your client app in the OS/firewall settings so it stops blocking you in the first place.

**Comment by NeedHelp4Indy** [comment]
The application listens to one port, this port does not vary. And I think this is the problem. Only one port is closed, now the question is where this happens. Normally the app runs on port 1090, after high load generation it is simply closed (see screen). 
![image](https://github.com/IndySockets/Indy/assets/131161379/89821844-ebcb-41a3-a57b-dc28353e9e0a)

Do you definitely close the application and one should configure that on operating system level?
I'll analyze if we have any influence on this programmatically.

**Comment by rlebeau** [comment]
> The application listens to one port, this port does not vary.

So, the application you are having trouble with is the **server**, not a **client**?

> And I think this is the problem.

A server listening on one port is not a problem at all (that is how most servers work).

> Do you definitely close the application and one should configure that on operating system level? I'll analyze if we have any influence on this programmatically.

I don't understand what you are asking.  But it is absolutely clear that your problem is happening outside of Indy, so I'm inclined to close this ticket.

**Comment by NeedHelp4Indy** [comment]
Now I have (hopefully) finally identified the problem.

![image](https://github.com/IndySockets/Indy/assets/131161379/c17e82cf-15e1-45b4-9c99-f275e451bba7)


Our application runs into a deadlock. In the Delphi code I must have realized that my predecessors implemented a logic that should normally prevent deadlocks. 
![image](https://github.com/IndySockets/Indy/assets/131161379/8a2319ed-699d-465f-bb99-7906b9bde5f8)

![image](https://github.com/IndySockets/Indy/assets/131161379/3c452dae-4e26-4861-ae6d-810a70bbb1c3)


Currently I have doubts to adapt the logic so that no deadlocks can occur anymore. For this I probably lack some background knowledge.

Alternatively, do you know of a way to programmatically check for deadlocks and remove them if necessary? 
Or do you have to take the approach of correcting the logic to prevent deadlocks in the first place?

**Comment by NeedHelp4Indy** [comment]
The CriticalSection is another potential cause of errors. I am additionally trying to get some help internally, somehow the deadlocks must be able to be prevented.

**Comment by rlebeau** [comment]
> Alternatively, do you know of a way to programmatically check for deadlocks and remove them if necessary? Or do you have to take the approach of correcting the logic to prevent deadlocks in the first place?

You can't simply remove a deadlock once it has occurred.  You have to prevent it from happening in the first place.
